injectdll/api.py

```diff
@@ -0,0 +1,121 @@
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of pywinauto nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import six
+
+from .injector import Injector
+from .channel import Pipe
+from .actionlogger import ActionLogger
+
+# backend exit codes enum
+OK = 0
+PARSE_ERROR = 1
+UNSUPPORTED_ACTION = 2
+MISSING_PARAM = 3
+RUNTIME_ERROR = 4
+NOT_FOUND = 5
+UNSUPPORTED_TYPE = 6
+INVALID_VALUE = 7
+
+
+class InjectedBaseError(Exception):
+    """Base class for exceptions based on errors returned from injected DLL side"""
+
+
+class InjectedUnsupportedActionError(InjectedBaseError):
+    """The specified action is not supported"""
+
+
+class InjectedRuntimeError(InjectedBaseError):
+    """Runtime exception during code execution inside injected target"""
+
+
+class InjectedNotFoundError(InjectedBaseError):
+    """Requested item not found: control element, property, ..."""
+
+
+class Singleton(type):
+    """
+    Singleton metaclass implementation from StackOverflow
+
+    http://stackoverflow.com/q/6760685/3648361
+    """
+
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+@six.add_metaclass(Singleton)
+class ConnectionManager(object):
+    def __init__(self):
+        self._pipes = {}
+
+    def _get_pipe(self, pid):
+        if pid not in self._pipes:
+            self._pipes[pid] = self._create_pipe(pid)
+        return self._pipes[pid]
+
+    def _create_pipe(self, pid):
+        pipe_name = 'process_{}'.format(pid)
```

Comment:
If you want to change the name of pipe, you also need to change it on the backend side here:
https://github.com/pywinauto/injected/blob/a7eee0e74c5e4ba75a705b3afa5fa036ddc0bc5a/backends/dotnet/src/wpf/Main.cs#L16
Note: prefix `pywinauto_` looks better for me, because it is less ambiguous and avoids possible conflicts with other applications (for example, some other process may want to create pipe with name `process_{PID}`

Reply:
@vasily-v-ryabov suggested removing all mention of the pywinauto project.
Maybe I should use `injectdll_` instead, what do you think?

Reply:
The most important thing is to specify the same pipe name on the client and server sides, otherwise this code will not work :)